compiler/noirc_evaluator/src/ssa/ssa_gen/context.rs

```diff
@@ -435,6 +447,68 @@ impl<'a> FunctionContext<'a> {
         }
     }
 
+    /// Checks that multiplying two u128 doesn't overflow because both operands are greater or equal than 2^64.
+    /// If both are, then the result is surely greater or equal than 2^128 so it would overflow.
+    /// The operands can still overflow if just one of them is less than 2^64, but in that case the result
+    /// will be less than 2^192 so it fits in a Field value, and acir will check that it fits in a u128.
+    fn check_u128_mul_overflow(&mut self, lhs: ValueId, rhs: ValueId) {
+        let lhs_value = self.builder.current_function.dfg.get_numeric_constant(lhs);
+        let rhs_value = self.builder.current_function.dfg.get_numeric_constant(rhs);
+
+        // If lhs is less than 2^64 then the condition trivially holds.
+        if let Some(value) = lhs_value {
+            if value.to_u128() < (1_u128 << 64) {
+                return;
+            }
+        }
+
+        // Same goes for rhs
+        if let Some(value) = rhs_value {
+            if value.to_u128() < (1_u128 << 64) {
+                return;
+            }
+        }
+
+        let u128 = NumericType::unsigned(128);
+
+        // This mask keeps the high-end bits of a u128 value.
+        let mask = 0xffffffffffffffff0000000000000000_u128;
+        let mask = self.builder.numeric_constant(mask, u128);
+
+        let res = if lhs_value.is_some() && rhs_value.is_some() {
+            // If both values are known at compile time, at this point we know it overflows
+            self.builder.numeric_constant(FieldElement::one(), u128)
+        } else if lhs_value.is_some() {
+            // If only the left-hand side is known we just need to check that the right-hand side
+            // isn't greater than 2^64
+            self.builder.insert_binary(rhs, BinaryOp::And, mask)
+        } else if rhs_value.is_some() {
+            // Same goes for the other side
+            self.builder.insert_binary(lhs, BinaryOp::And, mask)
+        } else {
+            // Here we do:
+            //
+            // lhs_masked = lhs & mask
+            // rhs_masked = rhs & mask
+            // res = lhs_masked * rhs_masked
+            // assert_eq(res, 0, "attempt to multiply with overflow")
+            //
+            // which is equivalent to checking that either of them is less than 2^64
+
+            let lhs_masked = self.builder.insert_binary(lhs, BinaryOp::And, mask);
+            let rhs_masked = self.builder.insert_binary(rhs, BinaryOp::And, mask);
```

Comment:
Euclidean division should be more efficient here rather than bitwise operations.

Reply:
Done! Let me know if I got this right.